# Data Understanding

In [12]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [13]:
df = pd.read_csv(r"..\data\raw\online_retail_II.csv")

In [14]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [15]:
df.shape

(1067371, 8)

In [16]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='str')

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 136.6 MB


In [18]:
# Missing Value Analysis

missing = df.isnull().sum()

missing_perc = (missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing values': missing,
    'Percentage': missing_perc
})

missing_df.sort_values(
    by = 'Percentage',
    ascending=False
)

,Missing values,Percentage
Customer ID,243007,22.766873
Description,4382,0.410541
StockCode,0,0.000000
Invoice,0,0.000000
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Country,0,0.000000


In [19]:
# Duplicate Analysis

duplicates = df.duplicated().sum()

duplicates

np.int64(34335)

In [20]:
# Statistical Summary

df.describe()

,Quantity,Price,Customer ID
count,1.067371e+06,1.067371e+06,824364.000000
mean,9.938898e+00,4.649388e+00,15324.638504
std,1.727058e+02,1.235531e+02,1697.464450
min,-8.099500e+04,-5.359436e+04,12346.000000
25%,1.000000e+00,1.250000e+00,13975.000000
50%,3.000000e+00,2.100000e+00,15255.000000
75%,1.000000e+01,4.150000e+00,16797.000000
max,8.099500e+04,3.897000e+04,18287.000000


1. Dataset contains over 1 million transaction records.

2. Approximately 22.8% of records have missing Customer IDs.

3. Around 34,335 duplicate transactions exist.

4. InvoiceDate is stored as an object datatype and requires conversion to datetime.

5. The dataset contains missing product descriptions, although the percentage is relatively low.

6. Negative values are present in Quantity and Price columns, indicating potential returns, cancellations, or data quality issues.

7. Customer ID is stored as a float datatype and should be converted to an appropriate identifier type.

8. The dataset contains both numerical and categorical features suitable for customer analytics.